File: src/weaponsprocurement/gui/WimGuiModalLayout.java
package weaponsprocurement.gui;

final class WimGuiModalLayout {
    final float width;
    final float height;
    final float padding;
    final float sectionGap;
    final float headingHeight;
    final float footerHeight;
    final float rowHeight;
    final float rowGap;
    final float listInset;

    WimGuiModalLayout(float width,
                      float height,
                      float padding,
                      float sectionGap,
                      float headingHeight,
                      float footerHeight,
                      float rowHeight,
                      float rowGap,
                      float listInset) {
        this.width = width;
        this.height = height;
        this.padding = padding;
        this.sectionGap = sectionGap;
        this.headingHeight = headingHeight;
        this.footerHeight = footerHeight;
        this.rowHeight = rowHeight;
        this.rowGap = rowGap;
        this.listInset = listInset;
    }

    float bodyTop() {
        return padding + headingHeight + sectionGap;
    }

    float headingTop() {
        return padding;
    }

    float actionRowY(float headerContentHeight, float gapAfterHeader) {
        return headingTop() + headerContentHeight + gapAfterHeader;
    }

    float footerButtonY(float buttonHeight) {
        return height - padding - buttonHeight;
    }

    float bodyHeight() {
        return Math.max(rowHeight, height - bodyTop() - sectionGap - footerHeight - padding);
    }

    float contentWidth() {
        return width - 2f * padding;
    }

    float listPanelHeight(int rowCount, boolean hasAbove, boolean hasBelow, float extraGapHeight) {
        int controlCount = Math.max(1, rowCount + (hasAbove ? 1 : 0) + (hasBelow ? 1 : 0));
        return verticalItemsHeight(controlCount, rowHeight, rowGap) + Math.max(0f, extraGapHeight) + 2f * listInset;
    }

    float verticalItemsHeight(int count, float itemHeight, float gap) {
        int safeCount = Math.max(1, count);
        return safeCount * itemHeight + Math.max(0, safeCount - 1) * gap;
    }
}

File: src/main/kotlin/weaponsprocurement/ui/WimGuiModalLayout.kt
package weaponsprocurement.ui

class WimGuiModalLayout(
    @JvmField val width: Float,
    @JvmField val height: Float,
    @JvmField val padding: Float,
    @JvmField val sectionGap: Float,
    @JvmField val headingHeight: Float,
    @JvmField val footerHeight: Float,
    @JvmField val rowHeight: Float,
    @JvmField val rowGap: Float,
    @JvmField val listInset: Float,
) {
    fun bodyTop(): Float = padding + headingHeight + sectionGap

    fun headingTop(): Float = padding

    fun actionRowY(headerContentHeight: Float, gapAfterHeader: Float): Float =
        headingTop() + headerContentHeight + gapAfterHeader

    fun footerButtonY(buttonHeight: Float): Float = height - padding - buttonHeight

    fun bodyHeight(): Float = Math.max(rowHeight, height - bodyTop() - sectionGap - footerHeight - padding)

    fun contentWidth(): Float = width - 2f * padding

    fun listPanelHeight(rowCount: Int, hasAbove: Boolean, hasBelow: Boolean, extraGapHeight: Float): Float {
        val controlCount = Math.max(1, rowCount + (if (hasAbove) 1 else 0) + (if (hasBelow) 1 else 0))
        return verticalItemsHeight(controlCount, rowHeight, rowGap) + Math.max(0f, extraGapHeight) + 2f * listInset
    }

    fun verticalItemsHeight(count: Int, itemHeight: Float, gap: Float): Float {
        val safeCount = Math.max(1, count)
        return safeCount * itemHeight + Math.max(0, safeCount - 1) * gap
    }
}
